### PR TITLE
Module 5

### PR DESCRIPTION
**Pull request by dan1dr**
### Sections for the notebook

1. **Define baseline**. Computed a baseline provided in the solution for comparing our model
2. **Model learning and generalization assesment**. We introduced the "pyramid" of metrics to abstract the problem and potential improvements.
3. **Feature Importance and Data Leakages**. We addressed the data leakages showed in class providing some additional context and visualization
4. **Improvements**. We implemented a new XGBoost model with some additional hypothesis
5. **Conclusions**. We described a bit about the findings we made along the notebook + additional improvements for our model

We took several parts from the solution provided in the class. With this, we tried to improve, redefine some functions and introduce some hypothesis in order to try to improve our initial model.

### Results achieved

Our model showed a high mean performance at **9.28%**, significantly surpassing the baseline at **2.22%**. However, its median performance at **1.14%** was lower than the baseline's **1.55%**, pointing to variability in performance across different scenarios.

### Overall Reflection

- **Metric Approach**: A dual metric approach was employed, merging high-level analysis (quarterly returns against benchmarks like the SP500) with detailed low-level analysis (variable importance, median performance), providing a well-rounded view of the model's capabilities and areas for improvement.
- **High-Risk, High-Reward Strategy**: The model’s inclination towards a high-risk, high-reward strategy was evident. It demonstrated the potential for significant gains in certain market conditions, although this comes with an increased risk profile. This aspect underscores the need for careful risk management and targeted improvements.
- **Potential Improvements Detected**: Identified areas for enhancement, such as optimizing temporal trends analysis, refining the top_n parameter for investment strategies, improving train-test splitting methods, and exploring alternative algorithms for stock selection. These areas offer avenues for fine-tuning the model's accuracy and adaptability.

The advancements in this notebook mark a significant step towards more sophisticated financial modeling, acknowledging that this is a very complex problem to model and optimize due to its unique characteristics. Financial markets are inherently volatile, influenced by lots of factors ranging from global economic trends to investor sentiment. This complexity is exaplained by the challenge of predicting future market behavior, needing not only a deep understanding of financial theories but also the ability to adapt to lots of factors.

### CODE REVIEW

File: src/module_5/module5.md [hunk 1]
```diff
@@ -0,0 +1,5880 @@
...
+        + coord_flip()
+        + scale_x_discrete(limits=feature_importance["feature"])
+    )
+    return plot
+```
+
+# Solution
+
+The solution propose will encompass the following structure:
+1. Define baseline: we will create a baseline model for performance comparison, which would be the 500 largest market cap tickers. We could even decide that    the SP500 itself might be the benchmark, but will use the same baseline as the proposed in the solution.
```
Comment (gsantam) [inline src/module_5/module5.md:1702]: > We could even decide that    the SP500 itself might be the benchmark

we could do this indeed, but this way we are creating our own "index" and we can adapt the code to equities which can't be compared against sp500

File: src/module_5/module5.md [hunk 2]
```diff
@@ -0,0 +1,5880 @@
...
+  </tbody>
+</table>
+</div>
+
+
+
+
+```python
+(
+    ggplot(
```
Comment (gsantam) [inline src/module_5/module5.md:2862]: how are you finding ggplot compared to seaborn or matplotlib? I think I told you this in another PR, but I recommend you to also try altair

File: src/module_5/module5.md [hunk 3]
```diff
@@ -0,0 +1,5880 @@
...
+    Median for model: 0.10471844503860958
+    Median for baseline: 0.015525563344158869
+    
+
+
+Our model does 38 pts better in average and the median is 9 points better, maybe too optimistic to be true. Let's explore what could be happening more deeply:
+
+## 2. Model learning and generalization assessment
+
+In machine learning, the ultimate goal for the model is to try to reduce the error function as much as possible. While this might be positive, sometimes the metric being optimized is not necessarily our best metric based on the business problem we are tackling. In this light, there exists a range of different metrics closer to the model or business, which might be interesting to keep in mind and be diligent to know in which area we are trying to optimize. We can denominate that "pyramid," referring to how close we are with respect to the model, e.g., low-level if we are closer to the model, or high-level if we abstract enough from the model and we are closer to the business.
```
Comment (gsantam) [inline src/module_5/module5.md:2910]: Very good explanation. Some people call the "low level" metrics "offline metrics" and the "high level" ones "online metrics"

File: src/module_5/module5.md [hunk 4]
```diff
@@ -0,0 +1,5880 @@
...
+
+- Low-Level Metrics (Closest to the Model):
```
Comment (gsantam) [inline src/module_5/module5.md:2912]: Important to keep in mind: low level metrics are differentiable / convex in most of it's points -> that's why they can be optimized by the algorithms (either by gradient descent like log loss or by convex optimization methods like SVM). Some other algorithms like xgboost require the losses to be twice differentiable (since they use a second degree Taylor expansion), I'm mentioning this since it's sometimes asked in interviews

If you think about it, mid / high level metrics aren't differentiable -> can't be directly optimized by the model

File: src/module_5/module5.md [hunk 5]
```diff
@@ -0,0 +1,5880 @@
...
+    - Accuracy: The proportion of correctly predicted instances. It's intuitive but can be misleading in imbalanced datasets.
+    - Precision and Recall: Precision measures the correctness achieved in positive prediction, while recall measures how many actual positives were correctly identified.
+    - F1-Score: The harmonic mean of precision and recall, useful when you need a balance between precision and recall.
+- High-Level Metrics (Direct Business Impact):
+    - Top 10 Weighted Return: Your primary metric, directly aligned with your business objective (portfolio performance).
+    - Sharpe Ratio: Measures risk-adjusted return; it could be a supplementary metric for evaluating the effectiveness of the investment strategy.
+
+For visualization, I've illustrated a basic graph the following:
+
+<img src="images/pyramid_metrics.png" width="800" height="550">
```
Comment (gsantam) [inline src/module_5/module5.md:2927]: very nice plot! I may steal it

File: src/module_5/module5.md [hunk 6]
```diff
@@ -0,0 +1,5880 @@
...
+    columns={"normalized_learning_training": "Normalized Learning"}, inplace=True
+)
+learning_rates_test.rename(
+    columns={"normalized_learning_valid_0": "Normalized Learning"}, inplace=True
+)
+
+learning_rates_train["Set"] = "Training"
+learning_rates_test["Set"] = "Validation"
+
+learning_rates_combined = pd.concat([learning_rates_train, learning_rates_test])
```
Comment (gsantam) [inline src/module_5/module5.md:3389]: If you concatenate at the beginning you don't need to do the same filters / modifications for train and test

File: src/module_5/module5.md [hunk 7]
```diff
@@ -0,0 +1,5880 @@
...
+
+    
+![png](module5_files/module5_72_0.png)
+    
+
+
+    
+
+
+Anormal importance given to close_0 (actual price). This means that my actual price will determine very strongly if I'll beat the index. We could argue that this might be somehow explained by momentum, but we saw in previous lessons that reverse stock splitting are causing an important data leakages for these cases.
```
Comment (gsantam) [inline src/module_5/module5.md:3637]: The momentum would make sense if we were comparing the price against the price of the same equity in another moment. Here is the absolute price which is somehow arbitrary

File: src/module_5/module5.md [hunk 8]
```diff
@@ -0,0 +1,5880 @@
...
+
+The reverse stock splitting in this case is leaking info to the past; i.e. if the company approved a stock split (bad signal) it means it didn't go under good circumstances and the price is being calculated backwards, so higher prices will be associated with potential future stock splits, hence company in not good economic position and performing bad against sp500.
+
+In this case, we were carrying an investigation to verify a different stock splitting event presented in class, but I wouldn't conclude that fast that there should be something wrong with the model and the importance given to close_0 value. If I had more time, I would like to research a bit and confirm if this might be also something related to the "regression to the mean" phenomenon, which could give us an idea on why lower stocks might have higher probabilities for surpassing sp500!
+
+One possible solution would be to have the adjusted-prices for stocks, so stock splitting would not affect actual prices, but we do not have that feature in the current dataset (maybe online), so we will just drop the suspicious columns.
```
Comment (gsantam) [inline src/module_5/module5.md:4328]: You mean to reverse the splitting? That would be a good idea, tbh, and I think I try it. However, I think at the end of the day the best is to delete price related features and normalize everything by the company market cap

File: src/module_5/module5.md [hunk 9]
```diff
@@ -0,0 +1,5880 @@
...
+    [16:49:55] WARNING: /workspace/src/c_api/c_api.cc:1240: Saving into deprecated binary model format, please consider using `json` or `ubj`. Model format will default to JSON in XGBoost 2.2 if not specified.
+
+
+
+    
+![png](module5_files/module5_100_1.png)
+    
+
+
+I understand that this horrible plot shows that no feature plays an important role here (x scale is ~0), as described previously.
```
Comment (gsantam) [inline src/module_5/module5.md:5478]: haha I also don't understand what is this, what C means here?

File: src/module_5/module5.md [hunk 10]
```diff
@@ -0,0 +1,5880 @@
...
+    
+    Median for model: 0.011363999999999999
+    Median for baseline: 0.015525563344158869
+    
+
+
+## Conclusions
+
+**Mean and Median Analysis: Model vs. Baseline**.
+The mean value of your model stands at approximately 9.28%, significantly higher than the baseline's mean of about 2.22%. This indicates that on average, your model predicts considerably higher returns compared to the baseline. However, when looking at the median values, a different picture emerges. The model's median is around 1.14%, lower than the baseline's 1.55%. This suggests that while the model excels in generating higher average returns, its performance across various scenarios is less consistent, as the median is more resistant to outliers than the mean.
```
Comment (gsantam) [inline src/module_5/module5.md:5860]: A possible explanation of this is that if you think about it the model is giving high probabilities to companies which are quite similar (or in other words, the returns are very correlated). This means that either all predicted companies go well or none of them do (so either very high or low returns). When doing classic portfolios, one tries to find companies with a) good returns b) are uncorrelated from each other (check out https://en.wikipedia.org/wiki/Modern_portfolio_theory). The model is doing a) but not b) so it would be quite interesting to think of a regularization term which penalized correlated things

File: src/module_5/module5.md [hunk 11]
```diff
@@ -0,0 +1,5880 @@
...
+
+
+1. **Model Performance Improvement**: Enhanced predictive capabilities with reduced validation error, showing improved performance compared to the `diff_ch_sp500_baseline`.
+2. **Risk Management and Variable Importance**: Increased robustness and reduced overfitting by addressing data leakage issues, leading to better risk handling by eliminating reliance on misleading variables.
+3. **Mean Performance Values**: Significant improvement in final mean values over the initial baseline, with the model outperforming the SP500 on average after refinements.
+4. **High-Risk, High-Reward Strategy**: Adoption of a strategy with higher risks but potentially higher rewards, capable of achieving exceptional results in certain periods, indicating a riskier approach.
+5. **Potential Areas for Future Improvement**:
```
Comment (gsantam) [inline src/module_5/module5.md:5869]: check my comment above

File: src/module_5/module5.md [hunk 12]
```diff
@@ -0,0 +1,5880 @@
...
+
+**High-level and low-level Metrics Approach**. In terms of metrics, we've adopted a dual approach. At the high level, we focused on overarching performance indicators like yearly returns compared to benchmarks like the SP500, assessing the general effectiveness of the model. At the low level, we concentrated on detailed aspects such as variable importance and median performance, which offer insights into the model's behavior in typical scenarios and its resilience to extreme cases. This combination of high-level and low-level analysis provides a comprehensive view, highlighting the model's strong average performance and areas for potential refinement.
+
+
+1. **Model Performance Improvement**: Enhanced predictive capabilities with reduced validation error, showing improved performance compared to the `diff_ch_sp500_baseline`.
```
Comment (gsantam) [inline src/module_5/module5.md:5865]: I'd also add: frame the problem as a ranking one and put constraints to chose uncorrelated portfolios

**Review by gsantam — APPROVED** [review]
